File: project/app/src/main/java/org/owntracks/android/location/MotionDetector.java
package org.owntracks.android.location;

import android.location.Location;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import javax.inject.Inject;

public class MotionDetector {
    private static final int MAX_STATIONARY_SECONDS = 300;

    private Location previousLocation = null;

    private boolean moving = false;

    private long previousMovingTime = 0;

    private MutableLiveData<Boolean> currentMotion = new MutableLiveData<Boolean>();

    @Inject
    public MotionDetector() {
    }

    public void onLocationChanged(Location location) {
        boolean newMoving = false;
        if (previousLocation != null) {
            float distance = location.distanceTo(previousLocation);
            if (distance > location.getAccuracy() && distance > previousLocation.getAccuracy()) {
                newMoving = true;
                previousMovingTime = location.getElapsedRealtimeNanos();
                previousLocation = location;
            } else if ((((location.getElapsedRealtimeNanos() - previousMovingTime)) / 1000000000) < MAX_STATIONARY_SECONDS) {
                newMoving = true;
            }
        } else {
            previousLocation = location;
        }
        if (moving != newMoving) {
            moving = newMoving;
            currentMotion.postValue(moving);
        }
    }

    public LiveData<Boolean> getCurrentMotion() {
        return currentMotion;
    }

    public boolean isMoving() {
        return moving;
    }
}

File: project/app/src/main/java/org/owntracks/android/location/MotionDetector.kt
package org.owntracks.android.location

import android.location.Location
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject
import javax.inject.Singleton

const val MAX_STATIONARY_SECONDS = 300

@Singleton
class MotionDetector @Inject constructor() {
  val currentMotion: StateFlow<Boolean>
    get() = mutableCurrentMotion

  private val mutableCurrentMotion = MutableStateFlow<Boolean>(false)

  private var previousLocation: Location? = null

  private var previousMovingTime: Long = 0

  private var moving = false

  suspend fun onLocationChanged(location: Location) {
    var newMoving: Boolean = false

    val localPreviousLocation = previousLocation

    if (localPreviousLocation != null) {
      val distance = location.distanceTo(localPreviousLocation)
      if (distance > location.accuracy && distance > localPreviousLocation.accuracy) {
        newMoving = true
        previousMovingTime = location.elapsedRealtimeNanos
        previousLocation = location
      } else if ((((location.elapsedRealtimeNanos - previousMovingTime)) / 1000000000) < MAX_STATIONARY_SECONDS) {
        newMoving = true
      }
    } else {
      previousLocation = location
    }
    if (moving !== newMoving) {
      moving = newMoving
      mutableCurrentMotion.emit(moving)
    }
  }

  fun getMoving(): Boolean {
    return moving
  }
}
